frontend/catalyst/python_interface/pass_api/pattern_rewriter.py

```diff
@@ -0,0 +1,936 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Pattern rewriter API for quantum compilation passes."""
+
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+from functools import singledispatchmethod
+from numbers import Number
+from typing import TypeAlias
+from uuid import UUID, uuid4
+
+from pennylane import math, measurements, ops
+from pennylane.exceptions import TransformError
+from pennylane.measurements import MeasurementProcess
+from pennylane.operation import Operator
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import arith, builtin, func, scf, tensor
+from xdsl.ir import Block, BlockArgument
+from xdsl.ir import Operation as xOperation
+from xdsl.ir import Region, SSAValue
+from xdsl.pattern_rewriter import PatternRewriter, PatternRewriterListener, PatternRewriteWalker
+from xdsl.rewriter import InsertPoint
+
+from catalyst.python_interface.dialects import quantum
+from catalyst.python_interface.utils import get_constant_from_ssa
+
+_named_observables = (ops.PauliX, ops.PauliY, ops.PauliZ, ops.Identity, ops.Hadamard)
+_gate_like_ops = (
+    quantum.CustomOp,
+    quantum.GlobalPhaseOp,
+    quantum.MultiRZOp,
+    quantum.PCPhaseOp,
+    quantum.QubitUnitaryOp,
+)
+
+
+@dataclass
+class DynamicWire:
+    """Dynamic wire label wrapper."""
+
+    var: SSAValue[builtin.i64] | None = None
+    """SSA value representing the wire's dynamic index into the quantum register."""
+
+    _id: UUID = field(default_factory=uuid4, init=False)
+    """Unique ID representing the qubit. This is only used if the wire was allocated dynamically
+    (``quantum.AllocQubitOp``) or the qubit originated as a ``BlockArgument``."""
+
+    def __hash__(self) -> int:
+        if self.var:
+            return hash(self.var)
+
+        return hash(self._id)
+
+    def __eq__(self, other: "DynamicWire") -> bool:
+        if self.var and other.var:
+            return self.var == other.var
+
+        return self._id == other._id
+
+
+WireLabelLike: TypeAlias = int | DynamicWire
+
+
+class PLPatternRewriter(PatternRewriter):
+    """A ``PatternRewriter`` with abstractions for quantum compilation passes.
+
+    This is a subclass of ``xdsl.pattern_rewriter.PatternRewriter`` that exposes
+    methods to abstract away low-level pattern-rewriting details relevant to
+    quantum compilation passes.
+    """
+
+    def __init__(self, current_operation: xOperation):
+        super().__init__(current_operation)
+
+    def get_qnode(
```

Comment:
I'm a bit confused by this one 🤔... I'm probably missing something. When would a user use this?